### ISSUE TITLE
[Idea]: Community map (maybe Show & Tell)

### ISSUE THREAD

**Issue by pjeweb** [alveusgg/alveusgg#668]
### Describe your idea / enhancement

It would be nice to have a map of the alveus community. This could be collected when posting show & tell posts. Original idea by Stolen87 in chat.

### How would this idea / enhancement be beneficial?

Community engagement, showing global reach of Alveus.

**Comment by MattIPv4** [comment]
If we're going to do this, I would recommend only allowing folks to select their nearest city, to avoid accidental/intentional doxxing.

**Comment by pjeweb** [comment]
> If we're going to do this, I would recommend only allowing folks to select their nearest city, to avoid accidental/intentional doxxing.

Yeah if not only the country more or less. Can round lat/lon to like 2 fractional digits or something like that.

**Comment by Stolers** [comment]
Hi, Stolen87 here (my public key is: trust me bro :D). I initially thought of just a short free text field that people can fill with whatever they want. Agree with @MattIPv4, true location data can be dangerous and it shouldn't be too precise.

**Comment by alvaroMorillas** [comment]
I would like to give this issue a try if you guys don't mind. I'm very new to this stack though and work is kinda hectic so it's going to take me a while to have something decent that lives up to what you guys do. If that's fine with you I will start working on it asap.

**Comment by pjeweb** [comment]
> I would like to give this issue a try if you guys don't mind. I'm very new to this stack though and work is kinda hectic so it's going to take me a while to have something decent that lives up to what you guys do. If that's fine with you I will start working on it asap.

Yeah for sure! If you don't mind, I think it would be best if you start with some sort of concept so you don't waste time for something we wont merge. Like features and maybe library choices, requirements for external services (maps) etc.

If you just want to mess around with some code thats fine too, but expect we might want some bigger changes in review ;)

**Comment by alvaroMorillas** [comment]
That's fine! If it doesn't appeal to you at least it will serve as training for me, I want to learn this stack anyways. 

I was thinking on creating a LocationPickerField component, so it can be reused if needed, and putting it on the ShowAndTellEntryForm. I found a SelectBoxField component already created but it doesn't quite work as I think it needs to for this use case, although I can be mistaken.

For the locations source I found a free to use (CC-BY License) and easy to set up API called GeoNames that has [A LOT](https://www.geonames.org/export/codes.html) of types of places (cities, mountains, lakes, buildings, etc.). We can filter it to just search for cities and villages so it's not too exact and possibly doxxes the user or Alveus if there's some mischievous person out there. With this API we have two options: first is use it as a regular API and call it on the Component. The usage rate it gives is plenty I think for this use case and to not spam the API and run out of credits I found a use-debounce library that works flawlessly, so we would be good on that front. The second option is to call the API with no query, get all places on planet Earth on a JSON and use that as our source. I tried that and this API offers 12.599.902 different locations. It would require some work on my part beforehand because some locations although technically in different coordinates are the same place and the API free tier max pagination is 5.000 results, but you guys have way more experience than me with this stuff so it would be your call on if its worth it to have it on a humongous local file rather than go to the API every time. I would share that script with you in case you wanted to update that list in the future, but I don't think we would be at risk of showing obsolete data unless something weird happened in the world.

The only question I have is what location are you guys interested in, I say this because the post location can be different from the user location if the user travelled, went on a hike or something like that. If you're looking for the post location it would be easy to just store the location name and coordinates in the post (ShowAndTellEntry) and call it a day. Maybe once the post is created the location can link to Google Maps in case anyone is interested in visiting. We would have to make sure to tell Maya to not open it on stream in case something leaks.

Storing the user location is possible too, but I don't know how to keep track of the users that create posts without logging in with Twitch. For them I would have to store the user location in ShowAndTellEntry instead of the User table, and that feels wrong to me, so maybe we only give that option to logged in users. I don't think this would give reliable data about Alveus demography anyways, but it would be cool to have a little flag next to your username if you want to disclose your location, something like "by pjeweb :country-flag:". I'll try that and if it looks too cheap or ancient I won't commit it. I would get the flags from FlagsAPI, also free to use (MIT License).

I still haven't searched too in depth for a way to draw a map with this data, but everything I find points me to Google APIs and I really don't want to deal with that, I'll get back to you if I find something nice. We could show user locations in a map, hours by location in another map, crossreference with data from other platforms... the possibilites are endless!

I'll probably require assistance with the libraries configuration though, just to make sure that all versions specified are good and don't break down the line with some random update, but that's pretty much my idea. Do you had anything like that in mind? Writing it in this post suddenly looks like way too much for such a simple request, maybe I got too excited.

**Comment by pjeweb** [comment]
> I was thinking on creating a LocationPickerField component, so it can be reused if needed, and putting it on the ShowAndTellEntryForm. I found a SelectBoxField component already created but it doesn't quite work as I think it needs to for this use case, although I can be mistaken.

That sounds great.

> For the locations source I found a free to use (CC-BY License) and easy to set up API called GeoNames that has [A LOT](https://www.geonames.org/export/codes.html) of types of places (cities, mountains, lakes, buildings, etc.). We can filter it to just search for cities and villages so it's not too exact and possibly doxxes the user or Alveus if there's some mischievous person out there. With this API we have two options: first is use it as a regular API and call it on the Component. The usage rate it gives is plenty I think for this use case and to not spam the API and run out of credits I found a use-debounce library that works flawlessly, so we would be good on that front. The second option is to call the API with no query, get all places on planet Earth on a JSON and use that as our source. I tried that and this API offers 12.599.902 different locations. It would require some work on my part beforehand because some locations although technically in different coordinates are the same place and the API free tier max pagination is 5.000 results, but you guys have way more experience than me with this stuff so it would be your call on if its worth it to have it on a humongous local file rather than go to the API every time. I would share that script with you in case you wanted to update that list in the future, but I don't think we would be at risk of showing obsolete data unless something weird happened in the world.

I wonder if we can just skip all the geocoding and just use a location selected on a map. So like maya said putting a pin on a map. This could also ask for the user device location and round the latitude/longitude.

If we want to do statistics on them you can still go through latitude/longitude pairs and map them to continents/countries etc.?

> The only question I have is what location are you guys interested in, I say this because the post location can be different from the user location if the user travelled, went on a hike or something like that. If you're looking for the post location it would be easy to just store the location name and coordinates in the post (ShowAndTellEntry) and call it a day. Maybe once the post is created the location can link to Google Maps in case anyone is interested in visiting. We would have to make sure to tell Maya to not open it on stream in case something leaks.

Yeah I think we'd be interesting in the post's location, not the submitter. We _could_  allow two locations (Im from… and Post is about …). Allowing even more sounds silly. People can just split it up into multiple posts if they do a world tour or something.

> Storing the user location is possible too, but I don't know how to keep track of the users that create posts without logging in with Twitch. For them I would have to store the user location in ShowAndTellEntry instead of the User table, and that feels wrong to me, so maybe we only give that option to logged in users. I don't think this would give reliable data about Alveus demography anyways, but it would be cool to have a little flag next to your username if you want to disclose your location, something like "by pjeweb :country-flag:". I'll try that and if it looks too cheap or ancient I won't commit it. I would get the flags from FlagsAPI, also free to use (MIT License).

Not sure if I agree. I think having it tied to the post is fine. We can still try to filter out similar post user names with the same location?

But I also like the idea of allowing people to add info about where they are from. Maybe this could on some sort of settings/profile page (see #86 #42).

> I still haven't searched too in depth for a way to draw a map with this data, but everything I find points me to Google APIs and I really don't want to deal with that, I'll get back to you if I find something nice. We could show user locations in a map, hours by location in another map, crossreference with data from other platforms... the possibilites are endless!

Yes!

> I'll probably require assistance with the libraries configuration though, just to make sure that all versions specified are good and don't break down the line with some random update, but that's pretty much my idea. Do you had anything like that in mind? Writing it in this post suddenly looks like way too much for such a simple request, maybe I got too excited.

I'd look into maplibre or leaflet to draw the pins for entries on the map and maybe look into clustering them. For posts per country and all you might need something more like a charting library that has country outlines.

Leaflet and maplibre should also work well for user input:
https://leafletjs.com/examples/quick-start/
https://maplibre.org/maplibre-gl-js/docs/examples/mouse-position/

**Comment by pjeweb** [comment]
The user profile location could be used to pre-fill the location when posting a show and tell post.

**Comment by pjeweb** [comment]
For the base map I always wanted to look into protomaps (https://protomaps.com/). But we can also use and proxy openstreemap.org or something. I think we could also afford a payed map service with the amount of users we have. But I'm not sure. Definitely would want to avoid google maps.

**Comment by fooji23** [comment]
Thank you for the resources, they all look great. Pinning on a map seems easy enough, although I would still put a textbox so people have the option to search for a place rather than play geoguessr and be frustrated because they can't seem to find that obscure place they went to on a map. Is accesibility something that I have to keep in mind? I'm a backend guy so I never have to worry about that.

I worked a little on this issue in May before life happened, and I was filtering locations in a way that was impossible for people to drop Alveus' address in a post. How can we work around that if we just let people pick any place on a map? Are all post approvers aware of said location and can delete it before it goes public? Should we just define a safe zone around Austin so nobody tries any funny business?

Other than that I think I have all that I need to start working on it. Sorry again for showing interest in this issue and then going silent for almost 6 months, I've been wanting to keep working on this for the longest time, just couldn't find any time for it.

Cheers!

**Comment by pjeweb** [comment]
> Thank you for the resources, they all look great. Pinning on a map seems easy enough, although I would still put a textbox so people have the option to search for a place rather than play geoguessr and be frustrated because they can't seem to find that obscure place they went to on a map. 

Sure that would be nice. It's just a matter of setting up the geocoding service. One of the biggest sources is OpenStreetMap. There is a free service we could use, but it has some limitations: https://operations.osmfoundation.org/policies/nominatim/. Setting up your own service is quite a bit of work. Otherwise we'd need to use a commercial service like Google, Mapbox or Maptiler. 

> Is accesibility something that I have to keep in mind? I'm a backend guy so I never have to worry about that.

Sure. We like to keep the app as accessible as possible. Using one of the established libraries should ensure there is at least decent keyboard controls. Having a text based alternative would be great too. But I would not focus too much on it or let yourself get discouraged by that.

> I worked a little on this issue in May before life happened, and I was filtering locations in a way that was impossible for people to drop Alveus' address in a post. How can we work around that if we just let people pick any place on a map? Are all post approvers aware of said location and can delete it before it goes public? Should we just define a safe zone around Austin so nobody tries any funny business?

I do not think we need anything special here. Everything get reviewed by trusted mods. I am sure we can spot malicious pins. Also I would suggest we round the coordinates to a level where you cant really get an exact address or even a particular neighbourhood. I would also limit the zoom for the maps (both input and display maps) to an appropriate level. One exception could be public places like zoos, if we care.

> Other than that I think I have all that I need to start working on it. Sorry again for showing interest in this issue and then going silent for almost 6 months, I've been wanting to keep working on this for the longest time, just couldn't find any time for it.

No worries at all! Happy you want to work on it!

**Comment by pjeweb** [comment]
Dohregard and DanMan went through 1700+ show and tell posts and extracted the location if possible. I am currently running them through a basic geocoder. That will only give us a point. @fooji23 I can get you a dump for testing if you want? I would just remove the `userId`. It would also not have any attachments ... But it might help to draw them on a map.

 It would probably be better to have a classification of continent/country/state/city/point or something like that?

**Comment by fooji23** [comment]
They're awesome for that, I was dreading going through all of them. A test dump would definitely help.

I pretty much finished the map component yesterday. I’m still stuck on a few concepts, so I’m a bit behind where I wanted to be by now, but I think it’ll look great once it’s all done.

> It would probably be better to have a classification of continent/country/state/city/point or something like that?

Not sure what you mean by that. Nominatim (the API I’m using for reverse search) already has [categories](https://nominatim.org/release-docs/develop/api/Output/#addressdetails) in "addressdetails." To avoid super-detailed addresses I’ve limited it to [municipality/city/town/village], [region/state/county], and [country]. If you want to tweak, add, or remove any levels, just let me know, it's easy to adjust.

I should have a PR ready for you in the next couple of weeks, assuming I don’t get slammed at work after vacation. Hopefully, we’ll have something ready for S&T soon. Sorry for the delays, hope I’m not holding things up too much.

Cheers!
